## NUKALA NAGA PAVAN





###    

#                                   Reading PISATest Scores

## Problem 1.1 - Dataset size

Load the training and testing sets and save them as variables with the names pisaTrain and pisaTest.

How many students are there in the training set?


In [6]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import seaborn as sn
%matplotlib inline 

In [7]:
pisaTrain = pd.read_csv("pisa2009train.csv")
pisaTest = pd.read_csv("pisa2009test.csv")

In [8]:
pisaTrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3663 entries, 0 to 3662
Data columns (total 24 columns):
grade                    3663 non-null int64
male                     3663 non-null int64
raceeth                  3628 non-null object
preschool                3607 non-null float64
expectBachelors          3601 non-null float64
motherHS                 3566 non-null float64
motherBachelors          3266 non-null float64
motherWork               3570 non-null float64
fatherHS                 3418 non-null float64
fatherBachelors          3094 non-null float64
fatherWork               3430 non-null float64
selfBornUS               3594 non-null float64
motherBornUS             3592 non-null float64
fatherBornUS             3550 non-null float64
englishAtHome            3592 non-null float64
computerForSchoolwork    3598 non-null float64
read30MinsADay           3629 non-null float64
minutesPerWeekEnglish    3477 non-null float64
studentsInEnglish        3414 non-null float64
schoo

In [9]:
pisaTrain.shape

(3663, 24)

By observing the above result , there are 3663 entries which are the number of students in the train dataset

## Problem 1.2 - Summarizing the dataset
What is the average reading test score of males & Of females??


In [11]:
pisaTrain[pisaTrain.male == 1].readingScore.mean() ## The average reading score of males in the train dataset

483.53247863247805

In [12]:
pisaTrain[pisaTrain.male == 0].readingScore.mean() ## The average reading score of females in the train dataset

512.94063093244

## Problem 1.3 - Locating missing values
Which variables have missing data in at least one observation in the training set? 

In [13]:
pisaTrain.isnull().sum()[ pisaTrain.isnull().sum() != 0]

raceeth                   35
preschool                 56
expectBachelors           62
motherHS                  97
motherBachelors          397
motherWork                93
fatherHS                 245
fatherBachelors          569
fatherWork               233
selfBornUS                69
motherBornUS              71
fatherBornUS             113
englishAtHome             71
computerForSchoolwork     65
read30MinsADay            34
minutesPerWeekEnglish    186
studentsInEnglish        249
schoolHasLibrary         143
schoolSize               162
dtype: int64




The above variables have missing values.


## Problem 1.4 - Removing missing values
Linear regression discards observations with missing data, so we will remove all such observations from the training and testing sets.


In [14]:
ptrain_nomiss = pisaTrain.dropna(axis=0)

In [15]:
ptrain_nomiss.isnull().sum()

grade                    0
male                     0
raceeth                  0
preschool                0
expectBachelors          0
motherHS                 0
motherBachelors          0
motherWork               0
fatherHS                 0
fatherBachelors          0
fatherWork               0
selfBornUS               0
motherBornUS             0
fatherBornUS             0
englishAtHome            0
computerForSchoolwork    0
read30MinsADay           0
minutesPerWeekEnglish    0
studentsInEnglish        0
schoolHasLibrary         0
publicSchool             0
urban                    0
schoolSize               0
readingScore             0
dtype: int64

In [16]:
ptest_nomiss = pisaTest.dropna(axis=0)

In [17]:
ptest_nomiss.isnull().sum()

grade                    0
male                     0
raceeth                  0
preschool                0
expectBachelors          0
motherHS                 0
motherBachelors          0
motherWork               0
fatherHS                 0
fatherBachelors          0
fatherWork               0
selfBornUS               0
motherBornUS             0
fatherBornUS             0
englishAtHome            0
computerForSchoolwork    0
read30MinsADay           0
minutesPerWeekEnglish    0
studentsInEnglish        0
schoolHasLibrary         0
publicSchool             0
urban                    0
schoolSize               0
readingScore             0
dtype: int64

## Problem 2.1 - Factor variables
Factor variables are variables that take on a discrete set of values, like the "States" of India. This is an unordered factor because there isn't any natural ordering between the states. An ordered factor has a natural ordering between the levels (an example would be the classifications "large," "medium," and "small").

1) Which of the following variables is an unordered factor with at least 3 levels? (Select all that apply.)

 grade
 
 male
 
 raceeth


In [18]:
ptrain_nomiss.grade.unique()

array([11, 10,  9, 12,  8], dtype=int64)

In [19]:
ptrain_nomiss.male.unique()

array([1, 0], dtype=int64)

In [20]:
ptrain_nomiss.raceeth.unique()

array(['White', 'Black', 'Hispanic', 'More than one race',
       'American Indian/Alaska Native', 'Asian',
       'Native Hawaiian/Other Pacific Islander'], dtype=object)

As , we can see that 'Grade' has 5 levels , but it's ordered. 'Male' is not ordered but it has only 2 levels.

'Raceeth' is the categorical variable with no order and more that 3 levels





##    

2)  Which of the following variables is an ordered factor with at least 3 levels? (Select all that apply.)
 
 grade
 
 male
 
 raceeth


In [21]:
ptrain_nomiss.raceeth.unique()

array(['White', 'Black', 'Hispanic', 'More than one race',
       'American Indian/Alaska Native', 'Asian',
       'Native Hawaiian/Other Pacific Islander'], dtype=object)

In [22]:
ptrain_nomiss.male.unique()

array([1, 0], dtype=int64)

In [23]:
ptrain_nomiss.grade.unique()

array([11, 10,  9, 12,  8], dtype=int64)

As , we can see that 'Raceeth' has 7 levels , but it's not ordered. 'Male' is not ordered but it has only 2 levels.

'Grade' is the categorical variable which is ordered and has more than 3 levels

## Problem 2.2 - Unordered factors in regression models
To include unordered factors in a linear regression model, we define one level as the "reference level" and add a binary variable for each of the remaining levels. In this way, a factor with n levels is replaced by n-1 binary variables. The reference level is typically selected to be the most frequently occurring level in the dataset.
As an example, consider the unordered factor variable "color", with levels "red", "green", and "blue". If "green" were the reference level, then we would add binary variables "colorred" and "colorblue" to a linear regression problem. All red examples would have colorred=1 and colorblue=0. All blue examples would have colorred=0 and colorblue=1. All green examples would have colorred=0 and colorblue=0.
Now, consider the variable "raceeth" in our problem, which has levels "American Indian/Alaska Native", "Asian", "Black", "Hispanic", "More than one race", "Native Hawaiian/Other Pacific Islander", and "White". Because it is the most common in our population, we will select White as the reference level.


Which binary variables will be included in the regression model? (Select all that apply.)

 raceethAmerican Indian/Alaska Native
 
 raceethAsian
 
 raceethBlack
 
 raceethHispanic
 
 raceethMore than one race
 
 raceethNative Hawaiian/Other Pacific Islander
 
 raceethWhite


 ## ANSWER:
 
 raceethAmerican Indian/Alaska Native

 raceethAsian 
 
 raceethBlack
 
 raceethHispanic
 
 raceethMore than one race
 
 raceethNative Hawaiian/Other Pacific Islander


The above mentioned will be binary variables.(except raceethWhite which is the reference variable and doesn't get included in the Linear Model as Binary variable)

## Problem 2.3 - Example unordered factors
Consider again adding our unordered factor race to the regression model with reference level "White".
For a student who is Asian, which binary variables would be set to 0? All remaining variables will be set to 1. (Select all that apply.)
 
 raceethAmerican Indian/Alaska Native                 -  0
 
 raceethAsian                                         -  1
 
 raceethBlack                                         -  0
 
 raceethHispanic                                      -  0
 
 raceethMore than one race                            -  0
 
 raceethNative Hawaiian/Other Pacific Islander        -  0
 

### For a student who is white, which binary variables would be set to 0? All remaining variables will be set to 1. (Select all that apply.)
 raceethAmerican Indian/Alaska Native                 -  0
 
 raceethAsian                                         -  0
 
 raceethBlack                                         -  0
 
 raceethHispanic                                      -  0
 
 raceethMore than one race                            -  0
 
 raceethNative Hawaiian/Other Pacific Islander        -  0


## Problem 3.1 - Building a model
Because the race variable takes on text values, it was loaded as a factor variable when we read in the dataset with read.csv() -- you can see this when you run str(pisaTrain) or str(pisaTest). However, by default R selects the first level alphabetically ("American Indian/Alaska Native") as the reference level of our factor instead of the most common level ("White"). Set the reference level of the factor. 

Now, build a linear regression model (call it lmScore) using the training set to predict readingScore using all the remaining variables.

It would be time-consuming to type all the variables, but R provides the shorthand notation "readingScore ~ ." to mean "predict readingScore using all the other variables in the data frame." The period is used to replace listing out all of the independent variables. As an example, if your dependent variable is called "Y", your independent variables are called "X1", "X2", and "X3", and your training data set is called "Train", instead of the regular notation:
What is the Multiple R-squared value of lmScore on the training set?


In [24]:
X_features = list(ptrain_nomiss.columns )
X_features.remove( 'readingScore' )
X_features

['grade',
 'male',
 'raceeth',
 'preschool',
 'expectBachelors',
 'motherHS',
 'motherBachelors',
 'motherWork',
 'fatherHS',
 'fatherBachelors',
 'fatherWork',
 'selfBornUS',
 'motherBornUS',
 'fatherBornUS',
 'englishAtHome',
 'computerForSchoolwork',
 'read30MinsADay',
 'minutesPerWeekEnglish',
 'studentsInEnglish',
 'schoolHasLibrary',
 'publicSchool',
 'urban',
 'schoolSize']

In [25]:
Train = pd.get_dummies(ptrain_nomiss[X_features], drop_first = True ) ## drop_first = True takes care of the reference level

In [26]:
Train.columns

Index(['grade', 'male', 'preschool', 'expectBachelors', 'motherHS',
       'motherBachelors', 'motherWork', 'fatherHS', 'fatherBachelors',
       'fatherWork', 'selfBornUS', 'motherBornUS', 'fatherBornUS',
       'englishAtHome', 'computerForSchoolwork', 'read30MinsADay',
       'minutesPerWeekEnglish', 'studentsInEnglish', 'schoolHasLibrary',
       'publicSchool', 'urban', 'schoolSize', 'raceeth_Asian', 'raceeth_Black',
       'raceeth_Hispanic', 'raceeth_More than one race',
       'raceeth_Native Hawaiian/Other Pacific Islander', 'raceeth_White'],
      dtype='object')

In [27]:
import statsmodels.api as sm

C:\Users\nagapavan\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [28]:
x = sm.add_constant(Train)

In [29]:
y = ptrain_nomiss.readingScore

In [30]:
lmScore = sm.OLS( y, x ).fit()

In [31]:
lmScore.params

const                                             76.489006
grade                                             29.542707
male                                             -14.521653
preschool                                         -4.463670
expectBachelors                                   55.267080
motherHS                                           6.058774
motherBachelors                                   12.638068
motherWork                                        -2.809101
fatherHS                                           4.018214
fatherBachelors                                   16.929755
fatherWork                                         5.842798
selfBornUS                                        -3.806278
motherBornUS                                      -8.798153
fatherBornUS                                       4.306994
englishAtHome                                      8.035685
computerForSchoolwork                             22.500232
read30MinsADay                          

In [32]:
lmScore.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           readingScore   R-squared:                       0.325
Model:                            OLS   Adj. R-squared:                  0.317
Method:                 Least Squares   F-statistic:                     41.04
Date:                Fri, 31 Aug 2018   Prob (F-statistic):          1.72e-180
Time:                        02:57:03   Log-Likelihood:                -13795.
No. Observations:                2414   AIC:                         2.765e+04
Df Residuals:                    2385   BIC:                         2.781e+04
Df Model:                          28                                         
Covariance Type:            nonrobust                                         
==================================================================================================================
                                                     coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------
const                                             76.4890     37.303      2.050      0.040       3.340     149.638
grade                                             29.5427      2.937     10.057      0.000      23.783      35.303
male                                             -14.5217      3.156     -4.601      0.000     -20.710      -8.333
preschool                                         -4.4637      3.486     -1.280      0.201     -11.300       2.372
expectBachelors                                   55.2671      4.294     12.871      0.000      46.847      63.687
motherHS                                           6.0588      6.091      0.995      0.320      -5.886      18.004
motherBachelors                                   12.6381      3.861      3.273      0.001       5.066      20.210
motherWork                                        -2.8091      3.522     -0.798      0.425      -9.715       4.097
fatherHS                                           4.0182      5.579      0.720      0.471      -6.923      14.959
fatherBachelors                                   16.9298      3.995      4.237      0.000       9.095      24.764
fatherWork                                         5.8428      4.396      1.329      0.184      -2.778      14.463
selfBornUS                                        -3.8063      7.324     -0.520      0.603     -18.168      10.555
motherBornUS                                      -8.7982      6.588     -1.336      0.182     -21.716       4.120
fatherBornUS                                       4.3070      6.264      0.688      0.492      -7.976      16.590
englishAtHome                                      8.0357      6.859      1.171      0.242      -5.415      21.487
computerForSchoolwork                             22.5002      5.703      3.946      0.000      11.318      33.683
read30MinsADay                                    34.8719      3.408     10.231      0.000      28.188      41.556
minutesPerWeekEnglish                              0.0128      0.011      1.194      0.233      -0.008       0.034
studentsInEnglish                                 -0.2866      0.228     -1.258      0.208      -0.733       0.160
schoolHasLibrary                                  12.2151      9.265      1.318      0.187      -5.953      30.383
publicSchool                                     -16.8575      6.726     -2.506      0.012     -30.046      -3.669
urban                                             -0.1101      3.963     -0.028      0.978      -7.881       7.661
schoolSize                                         0.0065      0.002      2.977      0.003       0.002       0.011
raceeth_Asian                                     63.1670     18.973      3.329      0.001   

##              From the lmScore model's summary, we obtained Multiple  R-squared:	0.325




##   

## Problem 3.2 - Computing the root-mean squared error of the model
What is the training-set root-mean squared error (RMSE) of lmScore?

## RMSE is obtained by actual 'y' values and predicted 'y' values

In [33]:
lmpredict = lmScore.predict( x )

In [34]:
lmpredict.head()

1    538.692128
3    495.626051
4    436.878125
7    566.133978
9    500.639739
dtype: float64

In [35]:
from statsmodels.tools.eval_measures import rmse

In [36]:
rmse( y, lmpredict )

73.365551432984503

## RMSE is  73.36

## Problem 3.3 - Comparing predictions for similar students
Consider two students A and B. They have all variable values the same, except that student A is in grade 11 and student B is in grade 9. What is the predicted reading score of student A minus the predicted reading score of student B?

 -59.09
 
 -29.54
 
 0
 
 29.54
 
 59.09
 
 The difference cannot be determined without more information about the two students


In [43]:
29.5427 *(11-9)  ## co-efficient of grade = 29.5427 and the values of grades are 11 , 9 . All other terms cancel each other.

59.0854

## The difference obtained will be 59.09

## Problem 3.4 - Interpreting model coefficients
What is the meaning of the coefficient associated with variable raceethAsian?

 Predicted average reading score of an Asian student
 
 Difference between the average reading score of an Asian student and the average reading score of a white student
 
 Difference between the average reading score of an Asian student and the average reading score of all the students in the dataset
 
 Predicted difference in the reading score between an Asian student and a white student who is otherwise identical


## ANSWER
 Difference between the average reading score of an Asian student and the average reading score of a white student
Because, the co-efficents determine the slope = (y2-y1)/(x2-x1) . Here 'x' is categorical.(denominator can't be zero)

## Problem 3.5 - Identifying variables lacking statistical significance
Based on the significance codes, which variables are candidates for removal from the model? Select all that apply. (We'll assume that the factor variable raceeth should only be removed if none of its levels are significant.)

 grade
 
 male
 
 raceeth
 
 preschool
 
 expectBachelors
 
 motherHS
 
 motherBachelors
 
 motherWork
 
 fatherHS
 
 fatherBachelors
 
 fatherWork
 
 selfBornUS
 
 motherBornUS
 
 fatherBornUS
 
 englishAtHome
 
 computerForSchoolwork
 
 read30MinsADay
 
 minutesPerWeekEnglish
 
 studentsInEnglish
 
 schoolHasLibrary
 
 publicSchool
 
 urban
 
 schoolSize


In [44]:
def get_insignificant_vars( lm ):
    var_p_vals_df = pd.DataFrame( lm.pvalues )
    var_p_vals_df['vars'] = var_p_vals_df.index
    var_p_vals_df.columns = ['pvals', 'vars']
    return list( var_p_vals_df[var_p_vals_df.pvals > 0.05]['vars'] )

In [45]:
insignificant_vars = get_insignificant_vars( lmScore)

In [46]:
insignificant_vars

['preschool',
 'motherHS',
 'motherWork',
 'fatherHS',
 'fatherWork',
 'selfBornUS',
 'motherBornUS',
 'fatherBornUS',
 'englishAtHome',
 'minutesPerWeekEnglish',
 'studentsInEnglish',
 'schoolHasLibrary',
 'urban',
 'raceeth_Black',
 'raceeth_Hispanic']

## Problem 4.1 - Predicting on unseen data
Using the "predict" function and supplying the "newdata" argument, use the lmScore model to predict the reading scores of students in pisaTest. Call this vector of predictions "predTest". Do not change the variables in the model (for example, do not remove variables that we found were not significant in the previous part of this problem). Use the summary function to describe the test set predictions.
What is the range between the maximum and minimum predicted reading score on the test set?


In [47]:
X_features = list(ptest_nomiss.columns )
X_features.remove( 'readingScore' )
X_features

['grade',
 'male',
 'raceeth',
 'preschool',
 'expectBachelors',
 'motherHS',
 'motherBachelors',
 'motherWork',
 'fatherHS',
 'fatherBachelors',
 'fatherWork',
 'selfBornUS',
 'motherBornUS',
 'fatherBornUS',
 'englishAtHome',
 'computerForSchoolwork',
 'read30MinsADay',
 'minutesPerWeekEnglish',
 'studentsInEnglish',
 'schoolHasLibrary',
 'publicSchool',
 'urban',
 'schoolSize']

In [53]:
Test = pd.get_dummies(ptest_nomiss[X_features], drop_first = True )

In [54]:
Test.columns

Index(['grade', 'male', 'preschool', 'expectBachelors', 'motherHS',
       'motherBachelors', 'motherWork', 'fatherHS', 'fatherBachelors',
       'fatherWork', 'selfBornUS', 'motherBornUS', 'fatherBornUS',
       'englishAtHome', 'computerForSchoolwork', 'read30MinsADay',
       'minutesPerWeekEnglish', 'studentsInEnglish', 'schoolHasLibrary',
       'publicSchool', 'urban', 'schoolSize', 'raceeth_Asian', 'raceeth_Black',
       'raceeth_Hispanic', 'raceeth_More than one race',
       'raceeth_Native Hawaiian/Other Pacific Islander', 'raceeth_White'],
      dtype='object')

In [55]:
x_test = sm.add_constant(Test)

In [56]:
y_test = ptest_nomiss.readingScore

In [57]:
x_test.shape

(990, 29)

In [58]:
lmpredict = lmScore.predict( x_test )

In [59]:
lmScore.params

const                                             76.489006
grade                                             29.542707
male                                             -14.521653
preschool                                         -4.463670
expectBachelors                                   55.267080
motherHS                                           6.058774
motherBachelors                                   12.638068
motherWork                                        -2.809101
fatherHS                                           4.018214
fatherBachelors                                   16.929755
fatherWork                                         5.842798
selfBornUS                                        -3.806278
motherBornUS                                      -8.798153
fatherBornUS                                       4.306994
englishAtHome                                      8.035685
computerForSchoolwork                             22.500232
read30MinsADay                          

In [80]:
lm_test_predict = lmScore.predict( x_test)

In [81]:
lm_test_predict.max() - lm_test_predict.min()

284.46831179513185

## Problem 4.2 - Test set SSE and RMSE
What is the sum of squared errors (SSE) of lmScore on the testing set?
What is the root-mean squared error (RMSE) of lmScore on the testing set?


In [74]:
class Stats:
    
    def __init__(self, X, y, model):
        self.data = X
        self.target = y
        self.model = model
        ## degrees of freedom population dep. variable variance
        self._dft = X.shape[0] - 1   
        ## degrees of freedom population error variance
        self._dfe = X.shape[0] - X.shape[1] - 1  
    
    def sse(self):
        '''returns sum of squared errors (model vs actual)'''
        squared_errors = (self.target - self.model.predict(self.data)) ** 2
        return np.sum(squared_errors)
        
    def sst(self):
        '''returns total sum of squared errors (actual vs avg(actual))'''
        avg_y = np.mean(self.target)
        squared_errors = (self.target - avg_y) ** 2
        return np.sum(squared_errors)
    
    def r_squared(self):
        '''returns calculated value of r^2'''
        return 1 - self.sse()/self.sst()
    
    def adj_r_squared(self):
        '''returns calculated value of adjusted r^2'''
        return 1 - (self.sse()/self._dfe) / (self.sst()/self._dft)

In [75]:
def pretty_print_stats(stats_obj):
    '''returns report of statistics for a given model object'''
    items = ( ('sse:', stats_obj.sse()), ('sst:', stats_obj.sst()), 
             ('r^2:', stats_obj.r_squared()), ('adj_r^2:', stats_obj.adj_r_squared()) )
    for item in items:
        print('{0:8} {1:.4f}'.format(item[0], item[1]))

In [82]:
stats = Stats(x_test, y_test, lmScore)
pretty_print_stats(stats)

sse:     5762082.3711
sst:     7798774.9638
r^2:     0.2612
adj_r^2: 0.2388


In [78]:
from sklearn import metrics

In [83]:
np.sqrt(metrics.mean_squared_error(lm_test_predict,y_test)) # RMSE 

76.290793831092344

## SSE = 5762082.37 , RMSE = 76.29

## Problem 4.3 - Baseline prediction and test-set SSE
What is the predicted test score used in the baseline model? Remember to compute this value using the training set and not the test set.
What is the sum of squared errors of the baseline model on the testing set? HINT: We call the sum of squared errors for the baseline model the total sum of squares (SST).


In [88]:
baseline =  sum(y) / len(y) ## here y - train dataset
baseline

517.96288732394294

In [90]:
base_test = sum(y_test) / len(y_test)
base_test

519.86427272727224

In [93]:
SSE_test = np.sum((y_test - base_test)**2)
SSE_test

7798774.963826365

## Problem 4.4 - Test-set R-squared
What is the test-set R-squared value of lmScore?


In [97]:
r_squared = (1-(5762082.37/SSE_test)) ## Since , R^2 = 1 - (SSE / SST)  , Here SSE=5762082.37 , SSE_test = SST
r_squared

0.261155451115503